File: app/src/main/java/com/example/jobdemo/MainActivityAdapter.java
package com.example.jobdemo;

import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.example.jobdemo.bean.ActivityBean;
import com.example.jobdemo.util.LogUtil;

import java.util.List;

/**
 * @author Administrator
 */
public class MainActivityAdapter extends RecyclerView.Adapter<MainActivityAdapter.MyViewHolder> {
    private final List<ActivityBean> activityS;

    public MainActivityAdapter(List<ActivityBean> activityS) {
        this.activityS = activityS;
    }


    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_mainactivity_rycyclerview, parent, false);
        return new MyViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
        holder.tv_demo_name.setText(activityS.get(position).getActivityChinesName());
        holder.itemView.setOnClickListener((v) -> {
            LogUtil.showD("activity名："+activityS.get(position).getClassName());
            v.getContext().startActivity(new Intent(v.getContext(), activityS.get(position).getClassName()));
        });
    }

    @Override
    public int getItemCount() {
        return activityS.size();
    }

    public static class MyViewHolder extends RecyclerView.ViewHolder {
        TextView tv_demo_name;
        CardView cv_root;

        public MyViewHolder(@NonNull View itemView) {
            super(itemView);
            tv_demo_name = itemView.findViewById(R.id.tv_demo_name);
            cv_root = itemView.findViewById(R.id.cv_root);
        }
    }
}

File: app/src/main/java/com/example/jobdemo/MainActivityAdapter.kt
package com.example.jobdemo

import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.jobdemo.bean.ActivityBean
import com.example.jobdemo.databinding.ItemMainactivityRycyclerviewBinding

/**
 * @author Administrator
 */
class MainActivityAdapter(private val activityS: List<ActivityBean>) :
    RecyclerView.Adapter<MainActivityAdapter.MyViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val binding = ItemMainactivityRycyclerviewBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false
        )
        return MyViewHolder(binding)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        holder.tvDemoName.text = activityS[position].activityChinesName
        holder.itemView.setOnClickListener { v: View ->
            v.context.startActivity(Intent(v.context, activityS[position].className))
        }
    }

    override fun getItemCount(): Int {
        return activityS.size
    }

    class MyViewHolder(binding: ItemMainactivityRycyclerviewBinding) :
        RecyclerView.ViewHolder(binding.root) {
        val tvDemoName: TextView = binding.tvDemoName
    }


    /**
     * 判断是否是每一组的头部
     */
    fun isGroupHead(position: Int): Boolean {
        return if (position == 0) {
            true
        } else {
            val currentGroup = getGroupName(position)
            val preGroup = getGroupName(position - 1)
            //判断当前名，与上一行是否一样，不一样就是头
            currentGroup != preGroup
        }
    }

    /**
     * 获取组名
     */
    fun getGroupName(position: Int): String = activityS[position].initial


    fun getSortLettersFirstPosition(letters: String?): Int {
        if (activityS.isEmpty()) {
            return -1
        }
        var position = -1
        for (index in activityS.indices) {
            if (activityS[index].initial.equals(letters)) {
                position = index
                break
            }
        }
        return position
    }
}